### PR TITLE
Resolve #92 add ability to create wallets thru RPC

### LINKED ISSUE #92: Add ability to create wallets through rpcwallet

### PR DESCRIPTION
**Pull request by hyc**
First draft, for review and comment only.

Open questions - should the CREATE command return the wallet seed to the client?
Should error messages from failed CREATE be returned to the client?
Should there be a command to list the wallets in the wallet_dir?

### CODE REVIEW

**Comment by NanoAkron** [comment]
Hmm...surely any transmission of the seed is highly susceptible to a MITM attack?

What use case can you imagine for needing to create a wallet on a remote machine?

**Comment by hyc** [comment]
I would expect remote usage to be thru TLS, using stunnel or similar.

It's not really about creating a wallet remotely. It's about being able to startup wallet-rpc and leave it running, so that things like LightWallet can just talk to it exclusively over RPC, instead of the current hack where the only way to create a wallet is by running monero-wallet-cli and parsing its stdout.

**Comment by NanoAkron** [comment]
Ah that sounds very good. 

Any chance we could daemonise the wallet like we do monerod?

**Comment by hyc** [comment]
Yeah I guess that could be added too.

The patch needed several tweaks but it's in running shape now.

File: src/wallet/wallet_rpc_server.cpp [hunk 1]
```diff
@@ -873,7 +933,8 @@ namespace tools
   //------------------------------------------------------------------------------------------------------------------------------
   bool wallet_rpc_server::on_query_key(const wallet_rpc::COMMAND_RPC_QUERY_KEY::request& req, wallet_rpc::COMMAND_RPC_QUERY_KEY::response& res, epee::json_rpc::error& er)
   {
-      if (m_wallet.restricted())
+    if (!m_wallet) return not_open(er);
+      if (m_wallet->restricted())
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server.cpp]: Indent

File: src/wallet/wallet_rpc_server.cpp [hunk 2]
```diff
@@ -1475,6 +1552,122 @@ namespace tools
...
+#ifdef _WIN32
+    if (!ptr)
+      ptr = strchr(req.filename.c_str(), '\\');
+#endif
+    if (ptr)
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+      er.message = "Invalid filename";
+      return false;
+    }
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server.cpp]: If the intent is to have a filename only, this probably needs more (ie, you could still get A:FOO on windows).

File: src/wallet/wallet_rpc_server.cpp [hunk 3]
```diff
@@ -1475,6 +1552,122 @@ namespace tools
...
+    std::string wallet_file = m_wallet_dir + "/" + req.filename;
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server.cpp]: This should probably use boost path functions with the / or /= operator

File: src/wallet/wallet_rpc_server.cpp [hunk 4]
```diff
@@ -1475,6 +1552,122 @@ namespace tools
...
+    std::unique_ptr<tools::wallet2> wal = tools::wallet2::make_new(vm2).first;
+    if (m_wallet)
+      delete m_wallet;
+    m_wallet = wal.release();
+    if (!m_wallet)
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+      er.message = "Failed to create wallet";
+      return false;
+    }
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server.cpp]: How about deleting m_wallet only after all checks return OK (and restore the pointer unless they do) ?

File: src/wallet/wallet_rpc_server.cpp [hunk 5]
```diff
@@ -1475,6 +1552,122 @@ namespace tools
...
+    hres.height = 0;
+    bool r = net_utils::invoke_http_json("/getheight", hreq, hres, m_http_client);
+    m_wallet->set_refresh_from_block_height(hres.height);
+    crypto::secret_key dummy_key;
+    try {
+      m_wallet->generate(wallet_file, req.password, dummy_key, false, false);
+    }
+    catch (const std::exception& e)
+    {
+      delete m_wallet;
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server.cpp]: m_wallet = NULL

File: src/wallet/wallet_rpc_server.cpp [hunk 6]
```diff
@@ -1475,6 +1552,122 @@ namespace tools
...
+      int argc = 3;
+      argv[0] = "wallet-rpc";
+      argv[1] = "--password";
+      argv[2] = req.password.c_str();
+      argv[3] = NULL;
+      vm2 = *m_vm;
+      command_line::add_arg(desc, arg_password);
+      po::store(po::parse_command_line(argc, argv, desc), vm2);
+    }
+    std::unique_ptr<tools::wallet2> wal = tools::wallet2::make_from_file(vm2, wallet_file).first;
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server.cpp]: Check whether this returns NULL and error out, keeping m_wallet if so

File: src/wallet/wallet_rpc_server_commands_defs.h [hunk 7]
```diff
@@ -909,5 +909,61 @@ namespace wallet_rpc
     };
   };
 
+  struct COMMAND_RPC_GET_LANGUAGES
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server_commands_defs.h]: Bump minor RPC version

File: src/wallet/wallet_rpc_server.cpp [hunk 8]
```diff
@@ -105,24 +113,56 @@ namespace tools
...
-    auto rpc_config = cryptonote::rpc_args::process(vm);
+    if (m_wallet)
+    {
+	  m_wallet->store();
+	  delete m_wallet;
+	  m_wallet = NULL;
+    }
+  }
+  //------------------------------------------------------------------------------------------------------------------------------
+  bool wallet_rpc_server::init(const boost::program_options::variables_map *vm)
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server.cpp]: Why change from reference to pointer ? I don't see anywhere obviously passing NULL.

File: src/wallet/wallet_rpc_server_commands_defs.h (hunk 7 above)
Comment (hyc) [inline src/wallet/wallet_rpc_server_commands_defs.h]: Wallet RPC has no versioning

Comment (NanoAkron) [inline src/wallet/wallet_rpc_server_commands_defs.h]: Should it be versioned?

Comment (hyc) [inline src/wallet/wallet_rpc_server_commands_defs.h]: Discussed this with @moneromooo-monero a week ago and decided it didn't need it.

Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server_commands_defs.h]: Yes, the version applies to the node RPC, and I just mixed up. Fine as is.

**Pull request merged**